mode/jsx/jsx.js

```diff
@@ -0,0 +1,217 @@
+// CodeMirror, copyright (c) by Marijn Haverbeke and others
+// Distributed under an MIT license: http://codemirror.net/LICENSE
+
+(function(mod) {
+  if (typeof exports == "object" && typeof module == "object") // CommonJS
+    mod(require("../../lib/codemirror"), require("../javascript/javascript"),
+        require("../xml/xml"));
+  else if (typeof define == "function" && define.amd) // AMD
+    define(["../../lib/codemirror", "../javascript/javascript", "../xml/xml"],
+           mod);
+  else // Plain browser env
+    mod(CodeMirror);
+})(function(CodeMirror) {
+  "use strict";
+
+  CodeMirror.defineMode("jsx", function(config, parserConfig) {
+    var tagOpen = /^\s*</,
+        closingTagOpen = /^\s*<\//,
+        modes = {
+          'js': CodeMirror.getMode(config, {
+            name: "javascript"
+          }),
+          'xml': CodeMirror.getMode(config, {
+            name: "xml",
+            htmlMode: true,
+            multilineTagIndentFactor: parserConfig.multilineTagIndentFactor,
+            multilineTagIndentPastTag: parserConfig.multilineTagIndentPastTag,
+            allowMissing: true
+          })
+        };
+
+    function tokenBase(stream, state) {
+      if (state.nextMode) {
+        state.currentMode = state.nextMode;
+        state.nextMode = null;
+      }
+
+      // If current mode is not XML and an XML tag opening gets detected next,
+      // switch editor mode to XML.
+      // ATTENTION: This should not happen inside an attribute JavaScript
+      // expression, since XML structures are not valid XML attributes.
+      if (state.currentMode != "xml" && !state.context.inAttrJSExpression) {
+        if (stream.match(tagOpen, false)) {
```

Comment:
The edge cases should be even less right now. I made it stricter on when should it start considering a "tag open" according to the `lastType` detected by the JavaScript mode.

Reference: https://github.com/sourcelair/CodeMirror/commit/4f18f9c7474c490766bafa5e37d08bd85ba1d4e7#diff-8fe55bb9634da263c3fa302cdbefb06cR43

Also, it would be really helpful here if I could access also the following information:

-  the exact token recognized in the `lastType` (e.g. if the `lastType == "operator"`, knowing if it was `=` or `-` could help making highlighting even more semantic).
- if there are any new lines between the last token recognized and the current one.